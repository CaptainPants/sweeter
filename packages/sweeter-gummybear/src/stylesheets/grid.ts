import {
    GlobalCssClass,
    GlobalCssStylesheet,
    StylesheetDependencyProvider,
    StylesheetFragmentBuilder,
    stylesheet,
} from '@captainpants/sweeter-web';
import { createConstantMap } from '../internal/createConstantMap.js';
import {
    breakpointNames,
    breakpointSizes,
    columnWidthIdentifiers,
    columnWidthNames,
} from './internal/constants.js';
import { themeStructure } from '../theme/themeStructure.js';

export const container = new GlobalCssClass({
    className: 'container',
    content: stylesheet`
        max-width: 900px;
        margin: 0 auto;
    `,
});

export const row = new GlobalCssClass({
    className: 'row',
    content: stylesheet`
        width: 100%;
        display: flex;
        flex-direction: row;
        flex-wrap: wrap;

        margin-left: calc(-1 * var(${themeStructure.grid.columnPadding.cssVar}));
        margin-right: calc(-1 * var(${themeStructure.grid.columnPadding.cssVar}));
        margin-bottom: var(${themeStructure.grid.rowBottomPadding.cssVar});
    `,
});

const dependencies = new StylesheetDependencyProvider();

export const columns = createConstantMap(
    breakpointNames,
    (breakpointName, breakpointIndex) => {
        const breakpointSize = breakpointSizes[breakpointIndex]!;

        if (breakpointSize === undefined) {
            return createConstantMap(
                columnWidthIdentifiers,
                (columnSizeName, i) => {
                    return new GlobalCssClass({
                        className: 'col-' + columnWidthNames[i],
                        extraDependencies: dependencies,
                    });
                },
            );
        }

        return createConstantMap(
            columnWidthIdentifiers,
            (columnSizeName, i) => {
                return new GlobalCssClass({
                    className:
                        'col-' + breakpointName + '-' + columnWidthNames[i],
                    extraDependencies: dependencies,
                });
            },
        );
    },
);

const gridStylesheetCss = new StylesheetFragmentBuilder();
const allColClasses: GlobalCssClass[] = [];

for (
    let breakpointIndex = 0;
    breakpointIndex < breakpointNames.length;
    ++breakpointIndex
) {
    const breakpointSize = breakpointSizes[breakpointIndex]!;
    const breakpointName = breakpointNames[breakpointIndex]!;

    const thisBreakpointCss = new StylesheetFragmentBuilder();

    for (
        let columnIndex = 0;
        columnIndex < columnWidthIdentifiers.length;
        ++columnIndex
    ) {
        const columnWidthIdentifier = columnWidthIdentifiers[columnIndex]!;

        const currentClass = columns[breakpointName][columnWidthIdentifier]!;
        allColClasses.push(currentClass);

        if (columnWidthIdentifier === 'auto') {
            // Auto-size
            thisBreakpointCss.appendLine(stylesheet`
                .${currentClass} {
                    flex: 1 1;
                }
            `);
        } else {
            thisBreakpointCss.appendLine(stylesheet`
                .${currentClass} {
                    flex: 0 0 ${((columnIndex + 1) * (100 / 12)).toFixed(4)}%;
                    width: 100%;
                    max-width: ${((columnIndex + 1) * (100 / 12)).toFixed(4)}%;
                }
            `);
        }
    }

    // No media query
    if (breakpointSize === undefined) {
        gridStylesheetCss.appendLine(thisBreakpointCss.build());
    }
    // Media query
    else {
        gridStylesheetCss.appendLine(stylesheet`
            @media screen and (min-width: ${breakpointSize}px) {
                ${thisBreakpointCss.build()}
            }
        `);
    }
}

// == Stuff thats common to all columns:
for (let i = 0; i < allColClasses.length; ++i) {
    if (i !== 0) {
        gridStylesheetCss.append(', ');
    }

    gridStylesheetCss.append(stylesheet`.${allColClasses[i]!}`);
}
gridStylesheetCss.appendLine(stylesheet`
    {
        padding-left: var(${themeStructure.grid.columnPadding.cssVar});
        padding-right: var(${themeStructure.grid.columnPadding.cssVar});
    }
`);
// == END OF Stuff thats common to all columns

// All the grid classes are defined in this stylesheet (so we have a circular reference here which is a bit hacky but its fine)
export const gridStylesheet = new GlobalCssStylesheet({
    id: 'grid',
    content: gridStylesheetCss.build(),
});

dependencies.addDependency(gridStylesheet).freeze();
