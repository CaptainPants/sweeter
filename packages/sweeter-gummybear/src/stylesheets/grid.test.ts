import { gridStylesheet } from "./grid.js"

it('Check for unintended changes', () => {
    const result = gridStylesheet.getContent({ getPrefixedClassName: class_ => class_.className });

    expect(result).toMatchSnapshot();
})