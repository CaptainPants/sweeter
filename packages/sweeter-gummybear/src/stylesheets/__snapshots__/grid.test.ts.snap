// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check for unintended changes 1`] = `
".col-1 {
    flex: 0 0 8.3333%;
    width: 100%;
    max-width: 8.3333%;
}
.col-2 {
    flex: 0 0 16.6667%;
    width: 100%;
    max-width: 16.6667%;
}
.col-3 {
    flex: 0 0 25.0000%;
    width: 100%;
    max-width: 25.0000%;
}
.col-4 {
    flex: 0 0 33.3333%;
    width: 100%;
    max-width: 33.3333%;
}
.col-5 {
    flex: 0 0 41.6667%;
    width: 100%;
    max-width: 41.6667%;
}
.col-6 {
    flex: 0 0 50.0000%;
    width: 100%;
    max-width: 50.0000%;
}
.col-7 {
    flex: 0 0 58.3333%;
    width: 100%;
    max-width: 58.3333%;
}
.col-8 {
    flex: 0 0 66.6667%;
    width: 100%;
    max-width: 66.6667%;
}
.col-9 {
    flex: 0 0 75.0000%;
    width: 100%;
    max-width: 75.0000%;
}
.col-10 {
    flex: 0 0 83.3333%;
    width: 100%;
    max-width: 83.3333%;
}
.col-11 {
    flex: 0 0 91.6667%;
    width: 100%;
    max-width: 91.6667%;
}
.col-12 {
    flex: 0 0 100.0000%;
    width: 100%;
    max-width: 100.0000%;
}
.col-auto {
    flex: 1 1;
}
@media screen and (min-width: 576px) {
    .col-sm-1 {
        flex: 0 0 8.3333%;
        width: 100%;
        max-width: 8.3333%;
    }
    .col-sm-2 {
        flex: 0 0 16.6667%;
        width: 100%;
        max-width: 16.6667%;
    }
    .col-sm-3 {
        flex: 0 0 25.0000%;
        width: 100%;
        max-width: 25.0000%;
    }
    .col-sm-4 {
        flex: 0 0 33.3333%;
        width: 100%;
        max-width: 33.3333%;
    }
    .col-sm-5 {
        flex: 0 0 41.6667%;
        width: 100%;
        max-width: 41.6667%;
    }
    .col-sm-6 {
        flex: 0 0 50.0000%;
        width: 100%;
        max-width: 50.0000%;
    }
    .col-sm-7 {
        flex: 0 0 58.3333%;
        width: 100%;
        max-width: 58.3333%;
    }
    .col-sm-8 {
        flex: 0 0 66.6667%;
        width: 100%;
        max-width: 66.6667%;
    }
    .col-sm-9 {
        flex: 0 0 75.0000%;
        width: 100%;
        max-width: 75.0000%;
    }
    .col-sm-10 {
        flex: 0 0 83.3333%;
        width: 100%;
        max-width: 83.3333%;
    }
    .col-sm-11 {
        flex: 0 0 91.6667%;
        width: 100%;
        max-width: 91.6667%;
    }
    .col-sm-12 {
        flex: 0 0 100.0000%;
        width: 100%;
        max-width: 100.0000%;
    }
    .col-sm-auto {
        flex: 1 1;
    }
}
@media screen and (min-width: 720px) {
    .col-md-1 {
        flex: 0 0 8.3333%;
        width: 100%;
        max-width: 8.3333%;
    }
    .col-md-2 {
        flex: 0 0 16.6667%;
        width: 100%;
        max-width: 16.6667%;
    }
    .col-md-3 {
        flex: 0 0 25.0000%;
        width: 100%;
        max-width: 25.0000%;
    }
    .col-md-4 {
        flex: 0 0 33.3333%;
        width: 100%;
        max-width: 33.3333%;
    }
    .col-md-5 {
        flex: 0 0 41.6667%;
        width: 100%;
        max-width: 41.6667%;
    }
    .col-md-6 {
        flex: 0 0 50.0000%;
        width: 100%;
        max-width: 50.0000%;
    }
    .col-md-7 {
        flex: 0 0 58.3333%;
        width: 100%;
        max-width: 58.3333%;
    }
    .col-md-8 {
        flex: 0 0 66.6667%;
        width: 100%;
        max-width: 66.6667%;
    }
    .col-md-9 {
        flex: 0 0 75.0000%;
        width: 100%;
        max-width: 75.0000%;
    }
    .col-md-10 {
        flex: 0 0 83.3333%;
        width: 100%;
        max-width: 83.3333%;
    }
    .col-md-11 {
        flex: 0 0 91.6667%;
        width: 100%;
        max-width: 91.6667%;
    }
    .col-md-12 {
        flex: 0 0 100.0000%;
        width: 100%;
        max-width: 100.0000%;
    }
    .col-md-auto {
        flex: 1 1;
    }
}
@media screen and (min-width: 960px) {
    .col-lg-1 {
        flex: 0 0 8.3333%;
        width: 100%;
        max-width: 8.3333%;
    }
    .col-lg-2 {
        flex: 0 0 16.6667%;
        width: 100%;
        max-width: 16.6667%;
    }
    .col-lg-3 {
        flex: 0 0 25.0000%;
        width: 100%;
        max-width: 25.0000%;
    }
    .col-lg-4 {
        flex: 0 0 33.3333%;
        width: 100%;
        max-width: 33.3333%;
    }
    .col-lg-5 {
        flex: 0 0 41.6667%;
        width: 100%;
        max-width: 41.6667%;
    }
    .col-lg-6 {
        flex: 0 0 50.0000%;
        width: 100%;
        max-width: 50.0000%;
    }
    .col-lg-7 {
        flex: 0 0 58.3333%;
        width: 100%;
        max-width: 58.3333%;
    }
    .col-lg-8 {
        flex: 0 0 66.6667%;
        width: 100%;
        max-width: 66.6667%;
    }
    .col-lg-9 {
        flex: 0 0 75.0000%;
        width: 100%;
        max-width: 75.0000%;
    }
    .col-lg-10 {
        flex: 0 0 83.3333%;
        width: 100%;
        max-width: 83.3333%;
    }
    .col-lg-11 {
        flex: 0 0 91.6667%;
        width: 100%;
        max-width: 91.6667%;
    }
    .col-lg-12 {
        flex: 0 0 100.0000%;
        width: 100%;
        max-width: 100.0000%;
    }
    .col-lg-auto {
        flex: 1 1;
    }
}
@media screen and (min-width: 1200px) {
    .col-xl-1 {
        flex: 0 0 8.3333%;
        width: 100%;
        max-width: 8.3333%;
    }
    .col-xl-2 {
        flex: 0 0 16.6667%;
        width: 100%;
        max-width: 16.6667%;
    }
    .col-xl-3 {
        flex: 0 0 25.0000%;
        width: 100%;
        max-width: 25.0000%;
    }
    .col-xl-4 {
        flex: 0 0 33.3333%;
        width: 100%;
        max-width: 33.3333%;
    }
    .col-xl-5 {
        flex: 0 0 41.6667%;
        width: 100%;
        max-width: 41.6667%;
    }
    .col-xl-6 {
        flex: 0 0 50.0000%;
        width: 100%;
        max-width: 50.0000%;
    }
    .col-xl-7 {
        flex: 0 0 58.3333%;
        width: 100%;
        max-width: 58.3333%;
    }
    .col-xl-8 {
        flex: 0 0 66.6667%;
        width: 100%;
        max-width: 66.6667%;
    }
    .col-xl-9 {
        flex: 0 0 75.0000%;
        width: 100%;
        max-width: 75.0000%;
    }
    .col-xl-10 {
        flex: 0 0 83.3333%;
        width: 100%;
        max-width: 83.3333%;
    }
    .col-xl-11 {
        flex: 0 0 91.6667%;
        width: 100%;
        max-width: 91.6667%;
    }
    .col-xl-12 {
        flex: 0 0 100.0000%;
        width: 100%;
        max-width: 100.0000%;
    }
    .col-xl-auto {
        flex: 1 1;
    }
}
.col-1, .col-2, .col-3, .col-4, .col-5, .col-6, .col-7, .col-8, .col-9, .col-10, .col-11, .col-12, .col-auto, .col-sm-1, .col-sm-2, .col-sm-3, .col-sm-4, .col-sm-5, .col-sm-6, .col-sm-7, .col-sm-8, .col-sm-9, .col-sm-10, .col-sm-11, .col-sm-12, .col-sm-auto, .col-md-1, .col-md-2, .col-md-3, .col-md-4, .col-md-5, .col-md-6, .col-md-7, .col-md-8, .col-md-9, .col-md-10, .col-md-11, .col-md-12, .col-md-auto, .col-lg-1, .col-lg-2, .col-lg-3, .col-lg-4, .col-lg-5, .col-lg-6, .col-lg-7, .col-lg-8, .col-lg-9, .col-lg-10, .col-lg-11, .col-lg-12, .col-lg-auto, .col-xl-1, .col-xl-2, .col-xl-3, .col-xl-4, .col-xl-5, .col-xl-6, .col-xl-7, .col-xl-8, .col-xl-9, .col-xl-10, .col-xl-11, .col-xl-12, .col-xl-auto {
    padding-left: var(--column-padding);
    padding-right: var(--column-padding);
}
"
`;
